orderer/common/cluster/comm.go

```diff
@@ -615,11 +624,13 @@ func (stream *Stream) operateWithTimeout(invoke StreamOperation) (*orderer.StepR
 
 	select {
 	case r := <-responseChan:
+		report(r.err)
```

Comment:
shouldn't you report only if there is an error?

Reply:
Well if err is nil, then it will report a nil. 

Keep in mind that the actual consumer of this mechanism, is the consensus (etcdraft) layer which actually waits for the result of this function, be it error or nil:


```
report := func(err error) {
		if err != nil {
			atomicErr.Store(err.Error())
			c.Metrics.ProposalFailures.Add(1)
		}
		close(sentChan) // <---- here we close the channel to signal to the goroutine which forward to the leader that we have a result (success or error)
	}

c.rpc.SendSubmit(lead, req, report)

select {
case <-sentChan: // <---- here we wait for the result
case <-c.doneC:
	return errors.Errorf("chain is stopped")
case <-timer.C:
	return errors.Errorf("timed out (%v) waiting on forwarding to %d", c.opts.RPCTimeout, lead)
}
```





    
    

    
        

      	}

    





    
    

    
        

      <span class="blob-code-inner blob-code-marker" data-code-marker="+"><br></span>